beacon-chain/state/stategen/replayer.go

```diff
@@ -0,0 +1,347 @@
+package stategen
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/dberr"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1/block"
+	"github.com/sirupsen/logrus"
+	"go.opencensus.io/trace"
+)
+
+var ErrFutureSlotRequested = errors.New("cannot replay to future slots")
+var ErrNoCanonicalBlockForSlot = errors.New("none of the blocks found in the db slot index are canonical")
+var ErrInvalidDBBlock = errors.New("invalid block found in database")
+
+// HistoryAccessor describes the minimum set of database methods needed to support the ReplayerBuilder
+type HistoryAccessor interface {
+	HighestSlotBlocksBelow(ctx context.Context, slot types.Slot) ([]block.SignedBeaconBlock, error)
+	GenesisBlock(ctx context.Context) (block.SignedBeaconBlock, error)
+	Block(ctx context.Context, blockRoot [32]byte) (block.SignedBeaconBlock, error)
+	StateOrError(ctx context.Context, blockRoot [32]byte) (state.BeaconState, error)
+}
+
+// CanonicalChecker is used by the Replayer to find the highest canonical block in the database <= the target slot
+// in practice this should be satisfied by a type that uses the fork choice store.
+type CanonicalChecker interface {
+	IsCanonical(ctx context.Context, blockRoot [32]byte) (bool, error)
+}
+
+// CurrentSlotter is used to ensure requestedSlot < currentSlot, to fail fast on requests to replay to future slots
+type CurrentSlotter interface {
+	CurrentSlot() types.Slot
+}
+
+// Replayer encapsulates database query and replay logic. It can be constructed via a StateReplayerBuilder
+type Replayer interface {
+	// ReplayBlocks replays the blocks the Replayer knows about based on Builder params
+	ReplayBlocks(ctx context.Context) (state.BeaconState, error)
+	// ReplayToSlot invokes ReplayBlocks under the hood,
+	// but then also runs process_slots to advance the state past the root or slot used in the builder.
+	// for example, if you wanted the state to be at the target slot, but only integrating blocks up to
+	// slot-1, you could request Builder.ForSlot(slot-1).ReplayToSlot(slot)
+	ReplayToSlot(ctx context.Context, target types.Slot) (state.BeaconState, error)
+}
+
+var _ Replayer = &stateReplayer{}
+
+type stateReplayer struct {
+	s           state.BeaconState
+	descendents []block.SignedBeaconBlock
+	blockSlot   types.Slot
```

Comment:
More direct is better?

```suggestion
	targetSlot   types.Slot
```

Reply:
going w/ `target`